src/finn/transformation/fpgadataflow/ip_cache.py

```diff
@@ -0,0 +1,642 @@
+    - `cache.update(model)`: Update the cache by adding synthesized IPs that are not
+        yet cached into the cache.
+    - `cache.get_key(op, model)`: Get the key (string) of the given custom op
+    - `cache.get_hash_hex(key)`: Get the hex representation of the hash of the given key.
+    - `cache.get_num_cached_ips(model)`: Get the number of cached IPs in the given model.
+    """
+
+    allowed_hashfuncs: Final[list[str]] = ["sha256", "sha512", "blake2s", "blake2b"]
+
+    def __init__(
+        self,
+        cache_dir: Path,
+        hashfunc: str,
+        hls_clk_period: float,
+        cache_hls_clk: bool,
+        fpgapart: str,
+        cache_fpgapart: bool,
+    ) -> None:
+        """Construct a new IPCache object.
+
+        Args:
+            cache_dir: The path of the cache directory.
+            hashfunc: The name of the hash function to be used.
+            hls_clk_period: HLS clock period in ns.
+            cache_hls_clk: Use the HLS clock as part of the key.
+            fpgapart: FPGA-part used for HLSSynth and PrepareIP.
+            cache_fpgapart: Use the fpgapart as part of the key.
+        """
+        self.cache_dir = cache_dir
+        self.cache_hls_clk = cache_hls_clk
+        self.cache_fpgapart = cache_fpgapart
+
+        # Used to check validity of cache directory names
+        if sys.platform != "win32":
+            self.max_hash_len = os.pathconf("/", "PC_NAME_MAX")
+            self.max_path_len = os.pathconf("/", "PC_PATH_MAX")
+        else:
+            # TODO: Implement filesystem checks
+            # 256 seems to be the default max path length under windows
+            self.max_hash_len = 256
+            self.max_path_len = 256
+
+        if not self.cache_dir.exists():
+            self.cache_dir.mkdir()
+        log.info(f"Opened cache handler. Cache directory: {self.cache_dir}")
+        if hashfunc not in dir(hashlib):
+            raise FINNConfigurationError(f"There is no hash function with the name {hashfunc}!")
+        if hashfunc not in self.allowed_hashfuncs:
+            raise FINNConfigurationError(
+                f"Hash function {hashfunc} not available for caching. "
+                f"Choose one of: {self.allowed_hashfuncs}"
+            )
+
+        self.hashfunc_name = hashfunc
+        self.hasher: Callable = getattr(hashlib, hashfunc)
+
+        # Prepare some always needed values
+        # FINN Commit
+        self.finn_commit = subprocess.run(
+            shlex.split("git rev-parse HEAD"),
+            text=True,
+            capture_output=True,
+            cwd=Path(__file__).parent,
+        ).stdout.strip()
+        self.finn_commit_time = subprocess.run(
+            shlex.split("git show --quiet --format=%ai"),
+            text=True,
+            capture_output=True,
+            cwd=Path(__file__).parent,
+        ).stdout.strip()
+        log.info(f"FINN Commit reads: {self.finn_commit} (authored at: {self.finn_commit_time})")
+
+        # FINN HLSLIB Commit
```

Comment:
Do we really need to hash the finn-hlslib commit separately? Should be pinned and thus depend strictly on the FINN+ commit.

Unless one is working with a special local version of the dependency of course, but then all bets are off anyways if we cache via commit hash and not changed files.

Or does this exist in preparation for a future relaxation where we no longer hash strictly on the FINN+ commit, but still on the finn-hlslib commit?

Reply:
Ok, we can leave it in for now. I was just thinking of minimizing lines of code ^^